core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -183,39 +183,38 @@ object Channel {
           Pull.eval {
             F.deferred[Unit].flatMap { waiting =>
               state
-                .modify { case State(values, size, ignorePreviousWaiting @ _, producers, closed) =>
-                  if (values.nonEmpty || producers.nonEmpty) {
-                    var unblock_ = F.unit
-                    var allValues_ = values
-
-                    producers.foreach { case (value, producer) =>
-                      allValues_ = allValues_ :+ value
-                      unblock_ = unblock_ >> producer.complete(()).void
-                    }
+                .modify {
+                  case prev @ State(values, size, ignorePreviousWaiting @ _, producers, closed) =>
+                    if (shouldEmit(prev)) (State(List.empty, 0, None, List.empty, closed), prev)
+                    else (State(values, size, waiting.some, producers, closed), prev)
+                }
+                .flatMap {
+                  case s @ State(values, size, ignorePreviousWaiting @ _, producers, closed) =>
+                    if (shouldEmit(s)) {
+                      var unblock_ = F.unit
+                      var allValues_ = values
+                      var size_ = size
 
-                    val unblock = unblock_
-                    val allValues = allValues_
+                      producers.foreach { case (value, producer) =>
+                        size_ += 1
+                        allValues_ = value :: allValues_
+                        unblock_ = unblock_ >> producer.complete(()).void
```

Comment:
On this line, I have noticed that the call to `producer.complete(())` itself is lazy (a lazy parameter of `>>`), so none of the `producer` objects would have this method called until the aggregated `F` is run, which means that the reference to the `producer` itself would be kept (in the lazy block) until then. Is this necessary?

Otherwise, would it perhaps help to write this as follows?
```Scala
unblock_ = unblock_ <* producer.complete(())
```

Reply:
I don't think we should change lazy `>>` to eager `<*`. Despite the fact that *usually* the amount of producers shouldn't be big, it's actually unbounded. Thus, I think it isn't actually safe to change to `*>` operator. Moreover `unblock` is executed right after it's created, so I don't think we'll gain from this change.